### PR TITLE
v4.0-pre1

### PR DESCRIPTION
**Pull request by jdabtieu**
- Add profile page
- Add global leaderboard
- Add global points and stats
- Update install for better data separation & compatibility with Docker
- Bug Fixes
  - CSRF not enforced on auth pages
  - hCaptcha config incorrect
  - Practice problem rejudge not working

Todo:
- [ ] Allow editing upload files
- [ ] Improve test coverage
- Update points properly if a problem is:
  - [x] deleted
  - [x] rejudged
  - ~~[ ] changed from draft to public~~
  - [x] points edited

### CODE REVIEW

**Comment by jdabtieu** [comment]
Deferring **Allow editing upload files** to v4.0-pre2

**Comment by slightlyskepticalpotat** [comment]
Does this include a migration script? @jdabtieu

**Comment by jdabtieu** [comment]
> Does this include a migration script? @jdabtieu

Yes

File: src/templates/auth/forgotpassword.html [hunk 1]
```diff
@@ -5,7 +5,11 @@
 {% block main %}
 <h1 class="text-center">Forgot Password</h1>
 <form method="post">
-    <input class="form-control mb-3" name="email" placeholder="Email Address" type="text" required>
+    <input type="hidden" name="csrf_token" value="{{ csrf_token() }}"/>
+    <div class="form-floating">
+        <input class="form-control mb-3" name="email" id="email" placeholder="Email Address" type="text" required>
```
Comment (slightlyskepticalpotat) [inline src/templates/auth/forgotpassword.html:10]: reformat this pls my eyes

File: src/templates/contest/create.html [hunk 2]
```diff
@@ -6,8 +6,14 @@
 {% block main %}
 <h1>Create Contest</h1>
 <form autocomplete="off" method="post">
-    <input class="form-control mb-3" name="contest_id" placeholder="Contest ID" required>
-    <input class="form-control mb-3" name="contest_name" placeholder="Contest Name" required>
+    <div class="form-floating">
+        <input class="form-control mb-3" name="contest_id" id="id" placeholder="Contest ID" required>
```
Comment (slightlyskepticalpotat) [inline src/templates/contest/create.html:10]: reformat

File: src/templates/contest/create_problem.html [hunk 3]
```diff
@@ -5,9 +5,15 @@
 
 {% block main %}
 <h1>Create Problem</h1>
-<form autocomplete="off" method="post" enctype=multipart/form-data>
-    <input class="form-control mb-3" id="id" name="id" placeholder="ID" required>
-    <input class="form-control mb-3" id="name" name="name" placeholder="Name" required>
+<form autocomplete="off" method="post" enctype="multipart/form-data">
+    <div class="form-floating">
+        <input class="form-control mb-3" type="text" id="id" name="id" placeholder="Problem ID" required>
```
Comment (slightlyskepticalpotat) [inline src/templates/contest/create_problem.html:10]: reformat

File: src/templates/contest/notify.html [hunk 4]
```diff
@@ -7,7 +7,10 @@
 <h1>Notify Participants</h1>
 <p>This will send an email to all participants in your contest. Use this feature with caution.</p>
 <form autocomplete="off" method="post" id="create" name="create">
-    <input class="form-control mb-3" type="text" name="subject" placeholder="Subject" required>
+    <div class="form-floating">
+        <input class="form-control mb-3" type="text" name="subject" id="subject" placeholder="Subject" required>
```
Comment (slightlyskepticalpotat) [inline src/templates/contest/notify.html:11]: reformat

File: src/templates/layout.html [hunk 5]
```diff
@@ -29,6 +29,7 @@
                             <li><a class="nav-link" href="/">Home</a></li>
                             <li><a class="nav-link" href="/problems">Practice</a></li>
                             <li><a class="nav-link" href="/contests">Contests</a></li>
+                            <li><a class="nav-link" href="/ranking">Ranking</a></li>
```
Comment (slightlyskepticalpotat) [inline src/templates/layout.html:32]: "Rankings" or "Users" is more correct

File: src/templates/problem/create.html [hunk 6]
```diff
@@ -5,33 +5,49 @@
...
+        <input class="form-control mb-3" type="text" id="name" name="name" placeholder="Problem Name" required>
+        <label for="name">Problem Name</label>
+    </div>
     <div style="display: flex;">
         <textarea class="form-control mb-3 md-preview"
                   id="description"
                   name="description"
-                  rows="20"
-                  placeholder="Description"
+                  style="height: 400px;"
```
Comment (slightlyskepticalpotat) [inline src/templates/problem/create.html:21]: inline css? surely not

**Review by slightlyskepticalpotat — CHANGES_REQUESTED** [review]
formatting is a mess, and i suggest more testing of the migrate script because i don't have it locally installed

however, after it's merged i can help write more tests to cover edge cases
